### PR TITLE
PWM DMA based RGB Underglow for STM32

### PR DESCRIPTION
**Pull request by zvecr**
## Description

Based off <https://github.com/joewa/WS2812-LED-Driver_ChibiOS>, this PR aims to implement ws2812 PWM support within the new ws2812 framework.

| | Status |
|-|-|
| f072 | ? |
| f103 | :heavy_check_mark: |
| f303 | :heavy_check_mark: |
| f401/f411 | :heavy_check_mark: |

#### Notes
1. I have added a non documented `WS2812_PWM_TARGET_PERIOD` option
    * the original code seems to use a 10x larger value, but does not work with the testing i've done so far
    * if required, board maintainers can use whatever value works for them

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by tzarc** [comment]
Could this leverage something similar to https://github.com/hubmartin/WS2812B_STM32F103/blob/master/src/ws2812b.c to decrease the memory requirements?

**Comment by zvecr** [comment]
Maybe, at the moment I am not too worried about memory usage as its in the same ball park as the SPI driver.

Just trying to remove stuff like `keyboards/matrix/noah/ws2812_f4.c` from the repo, the Apache stuff in the PR queue, and provide #6516/#6675.

**Comment by tzarc** [comment]
Working with L0xx as well.

File: drivers/arm/ws2812_pwm.c [hunk 1]
```diff
@@ -1 +1,207 @@
...
+                [WS2812_PWM_CHANNEL - 1] = {.mode = PWM_OUTPUT_ACTIVE_HIGH, .callback = NULL},  // Turn on the channel we care about
+            },
+        .cr2  = 0,
+        .dier = TIM_DIER_UDE,  // DMA on update event for next period
+    };
+    //#pragma GCC diagnostic pop  // Restore command-line warning options
+
+    // Configure DMA
+    // dmaInit(); // Joe added this
+    dmaStreamAllocate(WS2812_DMA_STREAM, 10, NULL, NULL);
```
Comment (tzarc) [inline drivers/arm/ws2812_pwm.c]: For `future`, this line was changed to:
```
dmaStreamAlloc(WS2812_DMA_STREAM - STM32_DMA1_STREAM1, 10, NULL, NULL);
```
...as dmaStreamAllocate() has been removed, and is now ID-based, not stream pointer based.

**Comment by drashna** [comment]
`__attribute__((weak))` ✅ 

Want to test it, but .... not up to doing so ATM

File: drivers/arm/ws2812_pwm.c (hunk 1 above)
Comment (tzarc) [inline drivers/arm/ws2812_pwm.c]: dmaStreamAlloc(WS2812_DMA_STREAM - STM32_DMA1_STREAM1, 10, NULL, NULL);

**Comment by tzarc** [comment]
Confirmed working with this applied over current post-future master, using L0xx.

File: drivers/arm/ws2812_pwm.c [hunk 2]
```diff
@@ -1 +1,207 @@
...
+#ifndef WS2812_DMA_STREAM
+#    define WS2812_DMA_STREAM STM32_DMA1_STREAM2  // DMA Stream for TIMx_UP
+#endif
+#ifndef WS2812_DMA_CHANNEL
+#    define WS2812_DMA_CHANNEL 2  // DMA Channel for TIMx_UP
+#endif
+
+#ifndef WS2812_PWM_TARGET_PERIOD
+//#    define WS2812_PWM_TARGET_PERIOD 800000 // Original code is 800k...?
+#    define WS2812_PWM_TARGET_PERIOD 80000  // TODO: work out why 10x less on f303/f4x1
```
Comment (yulei) [inline drivers/arm/ws2812_pwm.c]: I have used the same codes for the noah keyboard(stm32f11ce based board), and it should use 800K, not sure why it was modified to 80000.

File: drivers/arm/ws2812_pwm.c [hunk 3]
```diff
@@ -1 +1,207 @@
...
+#endif
+
+/* --- PRIVATE CONSTANTS ---------------------------------------------------- */
+
+#define WS2812_PWM_FREQUENCY (STM32_SYSCLK / 2)                             /**< Clock frequency of PWM, must be valid with respect to system clock! */
```
Comment (yulei) [inline drivers/arm/ws2812_pwm.c]: The timer's frequency depends on the specific APB1 or APB2's frequency, and which was decided by the APB divider defined on the mcuconf.h. If anyone tuned the value, the PWM frequency will be incorrect, so it better to make this MACRO overridable as the pwm driver and time channel selections.

File: drivers/arm/ws2812_pwm.c (hunk 2 above)
Comment (zvecr) [reply drivers/arm/ws2812_pwm.c]: doesnt work with 800K :man_shrugging:

File: drivers/arm/ws2812_pwm.c (hunk 3 above)
Comment (zvecr) [reply drivers/arm/ws2812_pwm.c]: Feel free to add an incremental PR, but as this has been merged its too late to change it within this PR.

File: drivers/arm/ws2812_pwm.c (hunk 2 above)
Comment (yulei) [reply drivers/arm/ws2812_pwm.c]: Maybe the testing system was running with HSI clock(8mHz, not the HSE at 8x9 on f303 or 8x12 on f411)?

**Pull request merged**